### PR TITLE
database: switch to ifstream, less code, more idiomatic C++

### PR DESCRIPTION
**Pull request by oharboe**
More idiomatic C++ means less manually written error handling and more reliance on exceptions.

This was inspired by seing fread() turning up and I think ifstream should be faster because as it is inlined and we're reading gigabytes.

I suspect, but don't have a way to measure precisely and accurately enough a 10% improvement: https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/issues/1001#issuecomment-1517339354

10% might not seem like much, but if the performance problems in the GUI can be addressed, 10% grows to become a substantial part of the load time. Let's say the rest of the load is made 5x faster, then 20% today becomes 50% of the load time.

### CODE REVIEW

**Comment by oharboe** [comment]
I suspect, but don't have a way to measure precisely and accurately enough a 10% improvement: https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/issues/1001#issuecomment-1517339354

File: src/odb/include/odb/dbStream.h [hunk 1]
```diff
@@ -441,7 +403,7 @@ class dbIStream
 
   void checkStream()
   {
-    int marker = ftell(_f);
+    uint64_t marker = _f.tellg();
```
[the remark this answers is not in the document]
Comment (oharboe) [reply src/odb/include/odb/dbStream.h:406]: @maliberty A configuration error of this linting, ZASSERT() is using this value.

[the remark this answers is not in the document]
Comment (oharboe) [reply src/odb/include/odb/dbStream.h:406]: @maliberty w.r.t. this local change: my plan is to rebase this pull request on top of the fixes after https://github.com/The-OpenROAD-Project/OpenROAD/issues/3225

**Comment by oharboe** [comment]
@maliberty good to go if the computer says yes.

**Review by maliberty — APPROVED** [review]
Do you plan to do dbOStream too?

**Comment by oharboe** [comment]
> Do you plan to do dbOStream too?

No, I am focusing on read performance. I think it should be done though, cant take long.

**Pull request merged**